ios/testmanagerd/xcuitestrunner.go

```diff
@@ -389,40 +338,70 @@ func runXUITestWithBundleIdsXcode15Ctx(
 		"ubiquitous test identifiers":              uint64(1),
 	}}
 
-	ideDaemonProxy2 := newDtxProxyWithConfig(conn2, testconfig)
+	ideDaemonProxy2 := newDtxProxyWithConfig(conn2, testconfig, testListener)
 	caps, err := ideDaemonProxy1.daemonConnection.initiateControlSessionWithCapabilities(localCaps)
 	if err != nil {
-		return err
+		return fmt.Errorf("runXUITestWithBundleIdsXcode15Ctx: cannot initiate a control session with capabilities: %w", err)
 	}
 	log.WithField("caps", caps).Info("got capabilities")
 	authorized, err := ideDaemonProxy2.daemonConnection.authorizeTestSessionWithProcessID(pid)
 	if err != nil {
-		return err
+		return fmt.Errorf("runXUITestWithBundleIdsXcode15Ctx: cannot authorize test session: %w", err)
 	}
 	log.WithField("authorized", authorized).Info("authorized")
 
 	err = ideDaemonProxy2.daemonConnection.initiateControlSession(pid, proto)
 	if err != nil {
-		return err
+		return fmt.Errorf("runXUITestWithBundleIdsXcode15Ctx: cannot initiate a control session: %w", err)
 	}
 	log.Debug("control session initiated")
 
 	ideInterfaceChannel := ideDaemonProxy1.dtxConnection.ForChannelRequest(ProxyDispatcher{id: "dtxproxy:XCTestDriverInterface:XCTestManager_IDEInterface"})
 
 	err = ideDaemonProxy1.daemonConnection.startExecutingTestPlanWithProtocolVersion(ideInterfaceChannel, proto)
 
-	time.Sleep(60 * time.Second)
+	if ctx != nil {
+		select {
+		case <-ctx.Done():
+			log.Infof("Killing test runner with pid %d ...", pid)
+			err = appserviceConn.KillProcess(int(pid))
+			if err != nil {
+				return fmt.Errorf("runXUITestWithBundleIdsXcode15Ctx: cannot kill test process: %w", err)
+			}
+			log.Info("Test runner killed with success")
+
+			if testListener != nil {
+				(*testListener).TestRunnerKilled()
+			}
+		}
+		return nil
+	}
+
+	<-ideDaemonProxy1.proxyDispatcher.closeChannel
+	log.Debugf("Killing test runner with pid %d ...", pid)
+	err = appserviceConn.KillProcess(int(pid))
+	if err != nil {
+		return fmt.Errorf("runXUITestWithBundleIdsXcode15Ctx: cannot kill test process: %w", err)
+	}
+	log.Debugf("Test runner killed with success")
+	if testListener != nil {
+		(*testListener).TestRunnerKilled()
+	}
+	var signal interface{}
+	ideDaemonProxy1.proxyDispatcher.closedChannel <- signal
```

Comment:
This looks unnecessary complicated for and I honestly don't understand what's going on here. I guess you want to wait for the test execution to finish? That could be done through the `testListener` as it should receive the signals about the current status and it should be able to tell if a test is finished or not

Reply:
I agree, it looks more complicated than it needs to be. When the new `testListener` part is removed, the remaining part entirely comes from the main branch. This has been the main way of blocking test runners. It does 2 things:

1. It blocks forever until test is finished
2. If provided a context, it blocks until context is canceled or test finishes, whichever is earlier.

Before this PR, there has been 2 places where this block of code you mark is duplicated. This PR adds a third duplication, which I'm not proud of.

I think we can abstract this and reuse the new abstraction for all testmanagerd implementations but I'd rather do it in a separate PR. This one is already large enough to grasp and such refactoring will be invasive. Is that okay?